### PR TITLE
Kev/sms messaging email model

### PR DESCRIPTION
**Pull request by knguyen142**
https://trello.com/c/9WmNXire/67-improve-reporting-for-e-messaging

##### SUMMARY
In order to do more advanced filters on the reports using emails, we need to store the email in a table. This adds a model and the corresponding migration file for emails following a similar approach as the sms table. 

The overlapping columns `domain`, `date`, `couch_recipient_doc_type`, `couch_recipient`, `messaging_subevent` are just copied from the SMS (log) model https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/sms/models.py#L75-L78.

This is groundwork that @calellowitz and I depend on for our tasks.

### CODE REVIEW

**Review by calellowitz — APPROVED** [review]

**Review by orangejenny — COMMENTED** [review]
Looks good, my only comment is that most of the `Email` fields should disallow nulls - probably all of them except maybe subject and body.

File: corehq/apps/sms/models.py
Comment (knguyen142) [inline corehq/apps/sms/models.py:2854]: Add new line.

**Comment by knguyen142** [comment]
> Looks good, my only comment is that most of the `Email` fields should disallow nulls - probably all of them except maybe subject and body.

They were copied and pasted from https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/sms/models.py#L75-L78, I thought it was unusual too but there might be a reason why they're nullable? I'm happy to change it either way.

**Comment by orangejenny** [comment]
Huh, weird. @millerdev do you know why so many of the `Log` fields are null? If not, I'd be inclined to make the new fields not-nullable, and then loosen restriction that if we find a reason in the future.

**Comment by mkangia** [comment]
> know why so many of the Log fields are null?
I guess it might have something to do with [moving data from couch to SQL](If not, I'd be inclined to make the new fields not-nullable, and then loosen restriction that if we find a reason in the future.) where these fields were not populated/well populated.

> If not, I'd be inclined to make the new fields not-nullable, and then loosen restriction that if we find a reason in the future.

+1. If we are making a new model and are sure that those values are always passed, should be good to begin with having them not null.

**Comment by mkangia** [comment]
I think this was already spoken about but just to understand this better,
(just a quick note is fine, can skip details)

Why is this new Email model not inherited from Log model which is what has been done for all other such models in the sms app?

Do we intend to track emails triggered from other places and not just from alerts that trigger other events like sms?

searching by just a part of email is really powerful, shows the users email addresses of lots of other users. Should we ensure they enter the exact email address only? 
And if that is possible (and assuming that one can only send emails to users in the system), 
can we just search for the `recipient_id` on `MessagingEvent` then, after finding the users using their email?

**Comment by knguyen142** [comment]
> Why is this new Email model not inherited from Log model which is what has been done for all other such models in the sms app?

I can answer this question. It's because the `Log` model has sms specific fields like `phone_number`. I think it should be possible to refactor the inheritance on the models and not affect any database schema?

And I will change the fields to non-nullable.

**Comment by knguyen142** [comment]
I removed the null fields and regenerated the migration https://github.com/dimagi/commcare-hq/pull/27788/commits/152ee38cf28e5c4444b671c9324733677eaa3181. Last check before merge?

**Review by orangejenny — APPROVED** [review]
looks good to me

**Comment by mkangia** [comment]
> I think it should be possible to refactor the inheritance on the models and not affect any database schema?

Sounds good to me. `Log` is an abstract class, so extracting the fields that are specific to SMS + Calls into a separate class (PhoneLog?) and inheriting from that for the classes that inherit from `Log` right now, and then inheriting Email from Log should do it.
I am not going to block on this, so feel free to go ahead, looks like you already have an approval from @calellowitz and @orangejenny .. but if they think its a worth change, then we should do it.

File: corehq/apps/sms/models.py
Comment (knguyen142) [inline corehq/apps/sms/models.py:2862]: Arggh, I missed this `null=True` because I wasn't looking at the first argument. Do you think it's worth changing, and if so, can I do a quick PR to modify it since this hasn't been run or the proper way of creating another migration file?

Comment (orangejenny) [reply corehq/apps/sms/models.py:2862]: I'd revert the other PR, then re-revert it and add a commit with the change

**Pull request merged**
